src/conn/pool/mod.rs

```diff
@@ -338,6 +338,75 @@ impl Pool {
         self.poll_new_conn_inner(false)
     }
 
+    /// Try to acquire a new connection under a spinlock checking that we don't
+    /// violate `exist < self.pool_constraints.max()`.
+    fn spin_acquire_conn(&mut self) -> Option<Result<Async<GetConn>>> {
+        let mut exist = self.inner.exist.load(atomic::Ordering::Acquire);
+
+        // Try to atomically swap the exist value while checking our condition.
+        loop {
+            if exist >= self.pool_constraints.max() {
+                return None;
+            }
+
+            let swapped =
+                self.inner
+                    .exist
+                    .compare_and_swap(exist, exist + 1, atomic::Ordering::AcqRel);
+
+            if exist == swapped {
+                break;
+            }
+
+            exist = swapped;
```

Comment:
This should probably include a call to [`spin_loop_hint`](https://doc.rust-lang.org/std/sync/atomic/fn.spin_loop_hint.html).

Reply:
I avoided this (and `yield_now`) since they are load and hardware-dependent. It's _probably_ a good idea but hard to say.